### PR TITLE
Add TypeScript messaging

### PR DESCRIPTION
**Pull request by rachfop**

### CODE REVIEW

File: docs/dev-guide/tscript/index.mdx [hunk 1]
```diff
@@ -58,9 +58,9 @@ The Foundations section of the Temporal Developer's guide covers the minimum set
 
 The Features section of the Temporal Developer's guide provides basic implementation guidance on how to use many of the development features available to Workflows and Activities in the Temporal Platform.
 
-- [How to develop with Signals](/dev-guide/typescript/features#signals): A Signal is a message sent to a running Workflow Execution
-- [How to develop with Queries](/dev-guide/typescript/features#queries): A Query is a synchronous operation that is used to get the state of a Workflow Execution.
-- [How to define Signals and Queries statically or dynamically](/dev-guide/typescript/features#static-and-dynamic-signals-and-queries): Each Workflow timeout controls the maximum duration of a different aspect of a Workflow Execution.
+- [How to develop with Signals](/dev-guide/typescript/messages#signals): A Signal is a message sent to a running Workflow Execution
```
Comment (angelazhou32) [inline docs/dev-guide/tscript/index.mdx:61]:
```suggestion
- [How to develop with Signals](/dev-guide/typescript/messages#signals): A Signal is a message sent to a running Workflow Execution.
```

File: docs/dev-guide/tscript/messages.mdx [hunk 2]
```diff
@@ -0,0 +1,440 @@
+---
+id: messages
+title: Messages
+sidebar_label: Messages
+description: Explore using Signals in Temporal TypeScript to send messages to Workflows, with details on defining, sending, and handling Signals, including customization options.
```
Comment (angelazhou32) [inline docs/dev-guide/tscript/messages.mdx:5]:
```suggestion
description: Explore using Signals in Temporal with the TypeScript SDK to send messages to Workflows, with details on defining, sending, and handling Signals, including customization options.
```

File: docs/dev-guide/tscript/messages.mdx [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+}
+```
+
+The Workflow dynamically reacts to external events without interrupting its execution flow.
+There can be only one handler for any Signal type; calling `setHandler` multiple times for the same Signal type overwrites the existing handler.
+For example, if you first set Handler A for a given Signal, and then later in the Workflow Definition set Handler B for the same Signal, then only Handler B will be invoked upon receiving this Signal.
+
+### Send a Signal from a Temporal Client {#send-signal-from-client}
+
+**How to send a Signal from a Temporal Client**
```
Comment (fairlydurable) [inline docs/dev-guide/tscript/messages.mdx:114]:
```suggestion
```

File: docs/dev-guide/tscript/messages.mdx [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+  });
+}
+```
+
+The Workflow dynamically reacts to external events without interrupting its execution flow.
+There can be only one handler for any Signal type; calling `setHandler` multiple times for the same Signal type overwrites the existing handler.
+For example, if you first set Handler A for a given Signal, and then later in the Workflow Definition set Handler B for the same Signal, then only Handler B will be invoked upon receiving this Signal.
+
+### Send a Signal from a Temporal Client {#send-signal-from-client}
+
```
Comment (fairlydurable) [inline docs/dev-guide/tscript/messages.mdx:113]:
```suggestion
```

File: docs/dev-guide/tscript/messages.mdx [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+      group.add(userId);
+    } else {
+      groups.set(groupId, new Set([userId]));
+    }
+  });
+}
+```
+
+The Workflow dynamically reacts to external events without interrupting its execution flow.
+There can be only one handler for any Signal type; calling `setHandler` multiple times for the same Signal type overwrites the existing handler.
```
Comment (angelazhou32) [inline docs/dev-guide/tscript/messages.mdx:109]:
```suggestion
There can be only one handler for any Signal type. Calling `setHandler` multiple times for the same Signal type overwrites the existing handler.
```

File: docs/dev-guide/tscript/messages.mdx [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+import { joinSignal } from './workflows';
+
+const client = new Client();
+
+const handle = client.workflow.getHandle('workflow-id-123');
+
+await handle.signal(joinSignal, { userId: 'user-1', groupId: 'group-1' });
+```
+
+The getHandle method is used to obtain a reference, or a "handle", to a specific Workflow Execution, which will be used to interact with that Workflow.
```
Comment (angelazhou32) [inline docs/dev-guide/tscript/messages.mdx:130]:
```suggestion
The `getHandle` method is used to obtain a reference, or a "handle", to a specific Workflow Execution, which will be used to interact with that Workflow.
```

File: docs/dev-guide/tscript/messages.mdx [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+For example, if you first set Handler A for a given Signal, and then later in the Workflow Definition set Handler B for the same Signal, then only Handler B will be invoked upon receiving this Signal.
+
+### Send a Signal from a Temporal Client {#send-signal-from-client}
+
+**How to send a Signal from a Temporal Client**
+
+When a Signal is sent successfully from the Temporal Client, the [WorkflowExecutionSignaled](/references/events#workflowexecutionsignaled) Event appears in the Event History of the Workflow that receives the Signal.
+
+<!--come back and review this and reword-->
+The [`WorkflowHandle.signal`](https://typescript.temporal.io/api/interfaces/client.WorkflowHandle#signal) method synchronously awaits the signal sending?
```
Comment (angelazhou32) [inline docs/dev-guide/tscript/messages.mdx:117]: Maybe move this line (line 119) to under the description of the handle (line 132) and write:

The [WorkflowHandle.signal](https://typescript.temporal.io/api/interfaces/client.WorkflowHandle#signal) method is used to send the Signal to the handle of the Workflow.

File: docs/dev-guide/tscript/messages.mdx [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+
+A Workflow can send a Signal to another Workflow, in which case it's called an _External Signal_.
+
+When an External Signal is sent:
+
+- A [SignalExternalWorkflowExecutionInitiated](/references/events#signalexternalworkflowexecutioninitiated) Event appears in the sender's Event History.
+- A [WorkflowExecutionSignaled](/references/events#workflowexecutionsignaled) Event appears in the recipient's Event History.
+
+<!--Reword for we voice-->
+Another reason we use the getExternalWorkflowHandle method is to ensure that Workflows remain deterministic.
```
Comment (angelazhou32) [inline docs/dev-guide/tscript/messages.mdx:142]: Another reason we use the `getExternalWorkflowHandle` method is to ensure that Workflows remain deterministic.

File: docs/dev-guide/tscript/messages.mdx [hunk 9]
```diff
@@ -0,0 +1,440 @@
...
+
+The getHandle method is used to obtain a reference, or a "handle", to a specific Workflow Execution, which will be used to interact with that Workflow.
+
+### Send a Signal from a Workflow {#send-signal-from-workflow}
+
+A Workflow can send a Signal to another Workflow, in which case it's called an _External Signal_.
+
+When an External Signal is sent:
+
+- A [SignalExternalWorkflowExecutionInitiated](/references/events#signalexternalworkflowexecutioninitiated) Event appears in the sender's Event History.
```
Comment (angelazhou32) [inline docs/dev-guide/tscript/messages.mdx:138]: We can add 'The`SignalExternalWorkflowExecutionInitiated` Event contains the Signal name as well as a Signal payload. The corresponding Command to this Event is `SignalExternalWorkflowExecution`.'

File: docs/dev-guide/tscript/messages.mdx [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+- A [WorkflowExecutionSignaled](/references/events#workflowexecutionsignaled) Event appears in the recipient's Event History.
```
Comment (angelazhou32) [inline docs/dev-guide/tscript/messages.mdx:139]: We can add 'This Event type indicates the Workflow has received a Signal Event. The Event type contains the Signal name as well as a Signal payload'.

File: docs/dev-guide/tscript/messages.mdx [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+
+<!--Reword for we voice-->
+Another reason we use the getExternalWorkflowHandle method is to ensure that Workflows remain deterministic.
+Recall that one aspect of deterministic Workflows means not directly making network calls from the Workflow.
+This means that developers cannot use a Temporal Client directly within the Workflow code to send Signals or start other Workflows.
+Instead, to communicate between Workflows, we use getExternalWorkflowHandle to both ensure that Workflows remain deterministic and also that these interactions are recorded as Events in the Workflow's Event History.
```
Comment (angelazhou32) [inline docs/dev-guide/tscript/messages.mdx:145]:
```suggestion
Instead, to communicate between Workflows, we use `getExternalWorkflowHandle` to both ensure that Workflows remain deterministic and also that these interactions are recorded as Events in the Workflow's Event History.
```

File: docs/dev-guide/tscript/messages.mdx [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+      group.add(userId);
+    } else {
+      groups.set(groupId, new Set([userId]));
+    }
+  });
+}
+```
+
+The Workflow dynamically reacts to external events without interrupting its execution flow.
+There can be only one handler for any Signal type. Calling `setHandler` multiple times for the same Signal type overwrites the existing handler.
```
Comment (fairlydurable) [inline docs/dev-guide/tscript/messages.mdx:109]:
```suggestion
There can be only one handler for any Signal type.
Calling `setHandler` multiple times for the same Signal type overwrites the existing handler.
```

File: docs/dev-guide/tscript/messages.mdx (hunk 9 above)
Comment (fairlydurable) [reply docs/dev-guide/tscript/messages.mdx:138]: Added

File: docs/dev-guide/tscript/messages.mdx (hunk 10 above)
Comment (fairlydurable) [reply docs/dev-guide/tscript/messages.mdx:139]: Added

File: docs/dev-guide/tscript/messages.mdx (hunk 11 above)
Comment (fairlydurable) [reply docs/dev-guide/tscript/messages.mdx:145]: Fixed

File: docs/dev-guide/tscript/messages.mdx [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+  const handle = getExternalWorkflowHandle('workflow-id-123');
+  await handle.signal(joinSignal, { userId: 'user-1', groupId: 'group-1' });
+}
+```
+
+### Signal-With-Start {#signal-with-start}
+
+In some cases, you might need to send a Signal, but aren't sure whether the Workflow Execution is running yet.
+The Signal-with-Start feature in Temporal provides a solution for this.
+Temporal provides a solution for this.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/dev-guide/tscript/messages.mdx:161]: Done.

File: docs/dev-guide/tscript/messages.mdx (hunk 12 above)
Comment (fairlydurable) [reply docs/dev-guide/tscript/messages.mdx:109]: Fixed.

File: docs/dev-guide/tscript/messages.mdx (hunk 8 above)
Comment (fairlydurable) [reply docs/dev-guide/tscript/messages.mdx:142]: Fixed, and we-voice removed here.

File: docs/dev-guide/tscript/messages.mdx (hunk 7 above)
Comment (fairlydurable) [reply docs/dev-guide/tscript/messages.mdx:117]: Updated

File: docs/dev-guide/tscript/messages.mdx [hunk 14]
```diff
@@ -0,0 +1,440 @@
...
+A Workflow can send a Signal to another Workflow, in which case it's called an _External Signal_.
+
+When an External Signal is sent:
+
+- A [SignalExternalWorkflowExecutionInitiated](/references/events#signalexternalworkflowexecutioninitiated) Event appears in the sender's Event History.
+ The `SignalExternalWorkflowExecutionInitiated` Event contains the Signal name as well as a Signal payload.
+ The corresponding Command to this Event is `SignalExternalWorkflowExecution`.
+- A [WorkflowExecutionSignaled](/references/events#workflowexecutionsignaled) Event appears in the recipient's Event History.
+  This Event type indicates the Workflow has received a Signal Event.
+  The Event type contains the Signal name as well as a Signal payload
```
Comment (angelazhou32) [inline docs/dev-guide/tscript/messages.mdx:143]:
```suggestion
The Event type contains the Signal name as well as a Signal payload.
```

**Review by angelazhou32 — APPROVED** [review]

**Pull request merged**